src/Storages/StorageS3.cpp

```diff
@@ -955,6 +955,53 @@ class StorageS3Sink : public SinkToStorage
     std::mutex cancel_mutex;
 };
 
+static SinkToStoragePtr createStorageS3Sink(const StorageS3::Configuration & query_configuration, const ContextPtr & local_context, const Block & sample_block, std::optional<FormatSettings> format_settings, const CompressionMethod compression_method, const String & bucket, const String & key)
+{
+    if (containsGlobs(key))
+        throw Exception(ErrorCodes::DATABASE_ACCESS_DENIED, "S3 key '{}' contains globs, so the table is in readonly mode", key);
+
+    String target_key = key;
+    bool truncate_in_insert = local_context->getSettingsRef().s3_truncate_on_insert;
+    bool object_already_exists = S3::objectExists(*query_configuration.client, bucket, target_key, query_configuration.url.version_id, query_configuration.request_settings);
+
+    if (!truncate_in_insert && object_already_exists)
+    {
+        if (local_context->getSettingsRef().s3_create_new_file_on_insert)
+        {
+            size_t suffix_num = 1;
+            auto pos = target_key.find_first_of('.');
+            String new_key;
+            do
+            {
+                new_key = target_key.substr(0, pos) + "." + std::to_string(suffix_num) + (pos == std::string::npos ? "" : target_key.substr(pos));
+                ++suffix_num;
+            }
+            while (S3::objectExists(*query_configuration.client, bucket, new_key, query_configuration.url.version_id, query_configuration.request_settings));
+
+            target_key = new_key;
```

Comment:
`02302_s3_file_pruning ` fails because in the new code we don't save new key in `configuration.keys`, so in the subsequent read from this engine we won't see the data in new files. For example:
```
create table test (number UInt64) engine=S3('<some_path>/data.csv');
insert into test select * from numbers(10);
insert into test select * from numbers(10, 10) settings s3_create_new_files_on_insert=1;
select count() from test;
```
And it should return `20` as now we have `2` files